runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -364,24 +369,50 @@ public void dispose() throws Exception {
       super.dispose();
       checkFinishBundleTimer.cancel(true);
     } finally {
+      if (bundleStarted) {
+        invokeFinishBundle();
+      }
       doFnInvoker.invokeTeardown();
     }
   }
 
   @Override
   public void close() throws Exception {
-    super.close();
-
-    // sanity check: these should have been flushed out by +Inf watermarks
-    if (!sideInputs.isEmpty() && nonKeyedStateInternals != null) {
-      BagState<WindowedValue<InputT>> pushedBack =
-          nonKeyedStateInternals.state(StateNamespaces.global(), pushedBackTag);
+    try {
 
-      Iterable<WindowedValue<InputT>> pushedBackContents = pushedBack.read();
-      if (pushedBackContents != null && !Iterables.isEmpty(pushedBackContents)) {
-        String pushedBackString = Joiner.on(",").join(pushedBackContents);
+      // This is our last change to block shutdown of this operator while
+      // there are still remaining processing-time timers. Flink will ignore pending
+      // processing-time timers when upstream operators have shut down and will also
+      // shut down this operator with pending processing-time timers.
```

Comment:
Please add a comment here describing the limitations of this approach (basically, typical cases of an unbounded SDF applied to a small bounded PCollection will not be fault-tolerant) and maybe link to the underlying Flink JIRA issue if there is one filed? If there isn't, would be good to file one for Flink; and either way, file one for Beam explaining that the support is limited.

Also would be good to log something.

Reply:
The Flink issue about checkpointing is https://issues.apache.org/jira/browse/FLINK-2491. This is not the reason for the special timer-related code here, though, so it's not the correct place to mention this here. The reason is that processing-time timers in Flink are not treated as data that should not be dropped and this is currently not seen as an issue by most people. (I'm working on getting that changed, but it's more of an ideological than an implementation issue.)